### PR TITLE
13576: Updates to JWE builder functionality

### PR DESCRIPTION
**Pull request by ayoho**
For issue #13576

### CODE REVIEW

**Comment by ayoho** [comment]
#run-libby-bot
#build
#spawn.fullfat.buckets=com.ibm.ws.security.jwt_fat.builder,com.ibm.ws.security.jwt_fat.consumer,com.ibm.ws.security.jwtsso_fat,com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat

**Comment by ayoho** [comment]
#run-libby-bot
#build
#spawn.fullfat.buckets=com.ibm.ws.security.jwt_fat.builder,com.ibm.ws.security.jwt_fat.consumer,com.ibm.ws.security.jwtsso_fat,com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat

File: dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops [hunk 1]
```diff
@@ -356,7 +354,8 @@ ERROR_BUILDING_SIGNED_JWE=CWWKS6060E: The [{0}] JWT builder cannot create a JSON
...
 ERROR_BUILDING_SIGNED_JWE.useraction=Check the error that is included in the message for more information.
 
+# Do not translate "JwtBuilder"
 # 0=Builder config ID, 1=keyManagementKeyAlias value, 2=Trust store config ID
 KEY_MANAGEMENT_KEY_NOT_FOUND=CWWKS6061E: The [{0}] JWT builder cannot create a JSON Web Encryption (JWE) token because a key management key cannot be found. They key management key alias is [{1}], and the configured trust store is [{2}].
-KEY_MANAGEMENT_KEY_NOT_FOUND.explanation=A key management key is a key that is used to encrypt the Content Encryption Key of a JWE token. A key must be specified so the JWT builder can create JWE tokens.
-KEY_MANAGEMENT_KEY_NOT_FOUND.useraction=Verify that a key with the alias that is specified in the message exists in the trust store that is specified in the message.
+KEY_MANAGEMENT_KEY_NOT_FOUND.explanation=A key management key is a key that is used to encrypt the Content Encryption Key of a JWE token. A key must be specified so the JWT builder can create JWE tokens. A key can be specified by using the configuration attributes that are specified in the message or through the JwtBuilder API.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops:360]: `A key must be specified so the JWT builder can create JWE tokens.`

In this sentence, does "key" refer to the key management key or the CEK? I think it means the key management key. If so, this sentence and the one that follows should specify. "key management key":

` A key management key must be specified so the JWT builder can create JWE tokens. A key management key can be specified by using the key alias and configured trust store values that are specified in the message or by using the JwtBuilder API.`

**Review by dmuelle — CHANGES_REQUESTED** [review]

**Comment by andrea-pichler** [comment]
Completed the requested L2 message review

**Comment by ayoho** [comment]
#run-libby-bot

File: dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops (hunk 1 above)
Comment (ayoho) [reply dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops]: Good suggestion - just pushed a new commit with your suggested changes.

File: dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/utils/JweCreator.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        String keyId = kidbuilder.buildKeyId(keyManagementKey);
+        if (keyId != null) {
+            jwe.setKeyIdHeaderValue(keyId);
+        }
+    }
+
+    void setJweHeaders(JsonWebEncryption jwe, BuilderImpl builder, JwtConfig jwtConfig) {
+        jwe.setAlgorithmHeaderValue(getKeyManagementKeyAlgorithm(builder, jwtConfig));
+        jwe.setEncryptionMethodHeaderParameter(getContentEncryptionAlgorithm(builder, jwtConfig));
+        jwe.setHeader("typ", "JOSE");
```
Comment (teddyjtorres) [inline dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/utils/JweCreator.java:86]: For reference only: This is to comply with section 4.1.11 https://tools.ietf.org/html/rfc7516#section-4.1.11 of the JWE specification.

**Review by dmuelle — APPROVED** [review]
LGTM

File: dev/com.ibm.ws.security.jwt_fat.builder/fat/src/com/ibm/ws/security/jwt/fat/builder/utils/BuilderHelpers.java [hunk 3]
```diff
@@ -81,7 +81,7 @@ public static JSONObject setDefaultClaimsWithEncryption(String builderId, String
         settings.remove(HeaderConstants.ALGORITHM);
         settings.put(HeaderConstants.ALGORITHM, KeyMgmtKeyAlg);
         settings.put(HeaderConstants.ENCRYPTION, contentEncryptAlg);
-        settings.put(HeaderConstants.TYPE, "JWT");
+        settings.put(HeaderConstants.TYPE, "JOSE");
         settings.put(HeaderConstants.CONTENT_TYPE, "jwt");
```
Comment (teddyjtorres) [inline dev/com.ibm.ws.security.jwt_fat.builder/fat/src/com/ibm/ws/security/jwt/fat/builder/utils/BuilderHelpers.java:85]: No changes needed now, but may need to be set to "JWT" if causing problem with legacy implementations per https://tools.ietf.org/html/rfc7519#section-5.1.

File: dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/utils/JweCreator.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        jwe.setHeader("cty", "jwt");
```
Comment (teddyjtorres) [inline dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/utils/JweCreator.java:87]: No changes needed now, but may need to be set to "JWT" if causing problem with legacy implementations per https://tools.ietf.org/html/rfc7519#section-5.1.

**Review by teddyjtorres — APPROVED** [review]

**Pull request merged**